### PR TITLE
Feat:[EC-231]  Track event date for sync messages

### PR DESCRIPTION
**Pull request by danielchirinda**

### CODE REVIEW

File: distribution/sender/routes/db-sync-route.xml [hunk 1]
```diff
@@ -17,6 +17,8 @@
                 #{body.setDateCreated(new java.util.Date())}
                 #{body.setMessageUuid(T(java.util.UUID).randomUUID().toString())}
                 #{body.setRequestUuid(getProperty('event').requestUuid)}
+                #{body.setEventDate(body.getDateCreated())}
```
Comment (wluyima) [inline distribution/sender/routes/db-sync-route.xml:20]: I don't these changes work, please read the ticket descriptions carefully, you should copy this date from the debezium event

Comment (wluyima) [reply distribution/sender/routes/db-sync-route.xml:20]: I also don't see where you have updated the route test

File: app-api/src/main/java/org/openmrs/eip/app/management/entity/SenderSyncMessage.java [hunk 2]
```diff
@@ -55,6 +55,10 @@ public enum SenderSyncMessageStatus {
 	@Access(AccessType.FIELD)
 	private Date dateSent;
 	
+	@Column(name = "event_date")
+	@Access(AccessType.FIELD)
```
Comment (wluyima) [inline app-api/src/main/java/org/openmrs/eip/app/management/entity/SenderSyncMessage.java:59]: It's pointless to set access to field, yet the property has a getter/setter, please remove @Access

File: webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.ts [hunk 3]
```diff
@@ -16,4 +16,7 @@ export class SenderSyncMessage extends BaseEntity {
 
 	dateSent?: string;
 
+	eventDate?: string;
+
+
```
Comment (wluyima) [inline webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.ts:21]: Where is the other UI change to show the value?

**Review by wluyima — CHANGES_REQUESTED** [review]
Added some comments

File: app-api/src/main/java/org/openmrs/eip/app/management/entity/SenderSyncMessage.java [hunk 4]
```diff
@@ -55,6 +55,9 @@ public enum SenderSyncMessageStatus {
 	@Access(AccessType.FIELD)
 	private Date dateSent;
 	
+	@Column(name = "event_date")
```
Comment (wluyima) [inline app-api/src/main/java/org/openmrs/eip/app/management/entity/SenderSyncMessage.java:58]: This needs to be a required value

File: app-api/src/main/resources/liquibase-sender.xml [hunk 5]
```diff
@@ -351,4 +351,17 @@
...
     
+    <changeSet author="dchirinda" id="20220929-1500">
+        <preConditions onFail="MARK_RAN">
+            <not>
+                <columnExists columnName="event_date" tableName="sender_sync_message" />
+            </not>
+        </preConditions>
+        <comment>Add event_date column to sender_sync_message table</comment>
+
+        <addColumn tableName="sender_sync_message">
```
Comment (wluyima) [inline app-api/src/main/resources/liquibase-sender.xml:362]: Needs to be required

File: app-api/src/test/java/org/openmrs/eip/app/route/sender/SenderSyncMsgReaderRouteTest.java [hunk 6]
```diff
@@ -73,6 +73,8 @@ public void shouldLoadSenderSyncMessagesSortedByDateCreatedAndCallTheMsgProcesso
 		assertEquals(msgCount, msgs.size());
 		assertTrue(msgs.get(0).getDateCreated().getTime() > (msgs.get(2).getDateCreated().getTime()));
 		assertTrue(msgs.get(1).getDateCreated().getTime() > (msgs.get(2).getDateCreated().getTime()));
+		assertTrue(msgs.get(0).getDateSent().getTime() < (msgs.get(2).getDateSent().getTime()));
```
Comment (wluyima) [inline app-api/src/test/java/org/openmrs/eip/app/route/sender/SenderSyncMsgReaderRouteTest.java:76]: Why did you add these? There is no requirement to sort the events by date sent like we do for date created

File: app-api/src/main/java/org/openmrs/eip/app/management/entity/SenderSyncMessage.java (hunk 4 above)
Comment (wluyima) [inline app-api/src/main/java/org/openmrs/eip/app/management/entity/SenderSyncMessage.java:58]: i think this should be required, and for existing rows at upgrade you will need to add a custom changeset to match the value of date_created of the same row

File: app-api/src/main/resources/liquibase-sender.xml (hunk 5 above)
Comment (wluyima) [inline app-api/src/main/resources/liquibase-sender.xml:362]: We should mark this as required

File: webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.ts (hunk 3 above)
Comment (wluyima) [reply webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.ts:21]: Did you see this comment?

**Comment by wluyima** [comment]
I don't see where you have updated existing tests to ensure this works as expected

**Review by wluyima — CHANGES_REQUESTED** [review]
I added more comments

File: app-api/src/main/java/org/openmrs/eip/app/management/entity/SenderSyncMessage.java (hunk 4 above)
Comment (danielchirinda) [reply app-api/src/main/java/org/openmrs/eip/app/management/entity/SenderSyncMessage.java:58]: it can be required, but creating a change set to update existing rows i think there no need to do that as we don't have that table in productions

File: webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.ts (hunk 3 above)
Comment (danielchirinda) [reply webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.ts:21]: I created a Tooltip for that as we don't have space on the table

File: webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.component.html [hunk 7]
```diff
@@ -13,7 +13,7 @@ <h5><span class="badge badge-dark" i18n="@@common-total-count">Total Count</span
 	</tr>
 	</thead>
 	<tbody>
-	<tr *ngFor="let m of syncMessages">
+	<tr *ngFor="let m of syncMessages" ngbTooltip="{{ 'Event Date: ' +  (m.eventDate | date:'yyyy-MM-dd HH:mm:ss')}}" [openDelay]="1000">
```
[the remark this answers is not in the document]
Comment (wluyima) [reply webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.component.html:16]: I'm starting to think for date fields, we should only display yyyy-MM-dd with the tooltop being the full date with time that we we create more room

File: app-api/src/main/java/org/openmrs/eip/app/management/entity/SenderSyncMessage.java [hunk 8]
```diff
@@ -55,6 +55,9 @@ public enum SenderSyncMessageStatus {
 	@Access(AccessType.FIELD)
 	private Date dateSent;
 	
+	@Column(name = "event_date", nullable = false)
```
Comment (wluyima) [inline app-api/src/main/java/org/openmrs/eip/app/management/entity/SenderSyncMessage.java:58]: You are missing @NotNull

File: app-api/src/main/resources/liquibase-sender.xml [hunk 9]
```diff
@@ -351,4 +351,19 @@
...
+            <column name="event_date" type="DATETIME(3)">
+            	<constraints nullable="false"/>
+            </column>
+        </addColumn>
+    </changeSet>
+
```
Comment (wluyima) [inline app-api/src/main/resources/liquibase-sender.xml:368]: Where is the changeset to update all existing rows at uprgade?

File: app-api/src/main/java/org/openmrs/eip/app/management/entity/SenderSyncMessage.java (hunk 4 above)
Comment (wluyima) [reply app-api/src/main/java/org/openmrs/eip/app/management/entity/SenderSyncMessage.java]: ok

**Review by wluyima — CHANGES_REQUESTED** [review]
Added more comments

File: distribution/sender/routes/db-sync-route.xml [hunk 10]
```diff
@@ -17,6 +17,7 @@
                 #{body.setDateCreated(new java.util.Date())}
                 #{body.setMessageUuid(T(java.util.UUID).randomUUID().toString())}
                 #{body.setRequestUuid(getProperty('event').requestUuid)}
+                #{body.setEventDate(getProperty('dbzmEvent').dateCreated)}
```
[the remark this answers is not in the document]
Comment (wluyima) [reply distribution/sender/routes/db-sync-route.xml:20]: Yes

File: webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.component.html (hunk 7 above)
[the remark this answers is not in the document]
Comment (danielchirinda) [reply webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.component.html:16]: That would be great and we could have space in the table, because by now i didn't succeed to provide internationalization to the tooltip.

File: app-api/src/main/resources/liquibase-sender.xml (hunk 5 above)
Comment (wluyima) [reply app-api/src/main/resources/liquibase-sender.xml:362]: Since this table is not in production, you should then add this column directly on the original changeset that creates the table

File: webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.component.html (hunk 7 above)
[the remark this answers is not in the document]
Comment (danielchirinda) [reply webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.component.html:16]: Can i make the change to display it at the table?

[the remark this answers is not in the document]
Comment (wluyima) [reply webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.component.html:16]: In theory, we should display these dates as yyyy-MM-dd and also include a tooltip that shows the full date as yyyy-MM-dd HH:mm:ss

File: app-api/src/test/java/org/openmrs/eip/app/route/sender/DbSyncRouteTest.java [hunk 11]
```diff
@@ -49,6 +57,8 @@ public void shouldCreateAndSaveTheSenderSyncMessageForTheEvent() {
 		assertEquals(requestUuid, msg.getRequestUuid());
 		assertNotNull(msg.getDateCreated());
 		assertNotNull(msg.getMessageUuid());
+		assertNull(msg.getDateSent());
+		assertNotNull(msg.getEventDate());
```
Comment (wluyima) [inline app-api/src/test/java/org/openmrs/eip/app/route/sender/DbSyncRouteTest.java:61]: You should check that debeziumEvent.dateCreated matches eventDate on the sync message

File: webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.component.html [hunk 12]
```diff
@@ -8,6 +8,7 @@ <h5><span class="badge badge-dark" i18n="@@common-total-count">Total Count</span
 		<th i18n="@@common-message-id">Message Id</th>
 		<th i18n="@@common-operation">Operation</th>
 		<th i18n="@@common-status">Status</th>
+		<th i18n="@@common-dateSent">Date Event</th>
```
Comment (wluyima) [inline webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.component.html:11]: This should be **Event Date** and not the other way round

File: webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.component.html [hunk 13]
```diff
@@ -27,8 +28,9 @@ <h5><span class="badge badge-dark" i18n="@@common-total-count">Total Count</span
 			<i class="bi-envelope-dash text-warning ui-cursor-pointer" [hidden]="m.status != 'NEW'" [ngbTooltip]="statusWaitingTemplate" [openDelay]="1000"></i>
 			<i class="bi-envelope-check-fill text-success ui-cursor-pointer" [hidden]="m.status != 'SENT'" [ngbTooltip]="statusSentTemplate" [openDelay]="1000"></i>
 		</td>
-		<td>{{m.dateCreated | date:'yyyy-MM-dd HH:mm:ss'}}</td>
-		<td>{{m.dateSent | date:'yyyy-MM-dd HH:mm:ss'}}</td>
+		<td container="body" ngbTooltip="{{m.eventDate | date:'yyyy-MM-dd HH:mm:ss'}}" [openDelay]="1000">{{m.eventDate | date:'yyyy-MM-dd'}}</td>
```
Comment (wluyima) [inline webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.component.html:31]: Why are you adding container attribute? It wasn't there before

Comment (wluyima) [reply webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.component.html:31]: I thought on the call we said for the tooltip we should consider using a more user readable date format, in fact even the data we should use a more friendly format, so for the date use angular's predefined formats **mediumDate** format and for the tooltip use **full** format, let me know what you think when you try these formats, I think they look cooler.

File: webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.component.html (hunk 12 above)
Comment (wluyima) [reply webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.component.html:11]: Did you add a Portuguese translation for this?

**Review by wluyima — CHANGES_REQUESTED** [review]
I added comments

File: webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.component.html (hunk 13 above)
Comment (danielchirinda) [reply webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.component.html:31]: Done

File: app-api/src/test/java/org/openmrs/eip/app/route/sender/DbSyncRouteTest.java [hunk 14]
```diff
@@ -32,10 +36,14 @@ public void shouldCreateAndSaveTheSenderSyncMessageForTheEvent() {
 		final String op = "c";
 		final String requestUuid = "some-request-uuid";
 		final boolean snapshot = true;
+	    DebeziumEvent  debeziumEvent = this.createDebeziumEvent(table, uuid, requestUuid, op);
+	    debeziumEvent.setDateCreated(new Date());
 		Event event = createEvent(table, null, uuid, op);
```
[the remark this answers is not in the document]
Comment (danielchirinda) [reply app-api/src/test/java/org/openmrs/eip/app/route/sender/DbSyncRouteTest.java]: Done

File: app-api/src/test/java/org/openmrs/eip/app/route/sender/DbSyncRouteTest.java [hunk 15]
```diff
@@ -2,53 +2,62 @@
...
+        assertEquals(table, msg.getTableName());
+        assertEquals(uuid, msg.getIdentifier());
+        assertEquals(op, msg.getOperation());
+        assertEquals(snapshot, msg.getSnapshot());
+        assertEquals(requestUuid, msg.getRequestUuid());
+        assertNotNull(msg.getDateCreated());
+        assertNotNull(msg.getMessageUuid());
+        assertNull(msg.getDateSent());
+        assertNotNull(msg.getEventDate());
+        assertEquals(msg.getEventDate().getTime(), debeziumEvent.getDateCreated().getTime());
```
Comment (wluyima) [inline app-api/src/test/java/org/openmrs/eip/app/route/sender/DbSyncRouteTest.java:60]: Can't you just compare the dates without calling getTime()

File: app-api/src/test/resources/mgt_sender_sync_message.sql [hunk 16]
```diff
@@ -1,5 +1,5 @@
-INSERT INTO sender_sync_message (id, table_name, identifier, operation, snapshot, message_uuid, request_uuid, status, date_created)
-VALUES (2, 'person', 'abfd940e-32dc-491f-8038-a8f3afe3e35b', 'c', 0, '26beb8bd-287c-47f2-9786-a7b98c933c04', null, 'NEW', '2020-06-28 00:00:00.001'),
-       (1, 'person', 'abfd940e-32dc-491f-8038-a8f3afe3e35b', 'u', 0, '16beb8bd-287c-47f2-9786-a7b98c933c04', null, 'NEW', '2020-06-28 00:00:00.001'),
-       (3, 'person', 'abfd940e-32dc-491f-8038-a8f3afe3e35b', 'u', 0, '36beb8bd-287c-47f2-9786-a7b98c933c04', null, 'NEW', '2020-06-28 00:00:00.000'),
-       (4, 'person', 'abfd940e-32dc-491f-8038-a8f3afe3e35b', 'u', 0, '46beb8bd-287c-47f2-9786-a7b98c933c04', null, 'SENT', '2020-06-28 00:00:00.000');
+INSERT INTO sender_sync_message (id, table_name, identifier, operation, snapshot, message_uuid, request_uuid, status, date_created, date_sent, event_date)
+VALUES (2, 'person', 'abfd940e-32dc-491f-8038-a8f3afe3e35b', 'c', 0, '26beb8bd-287c-47f2-9786-a7b98c933c04', null, 'NEW', '2020-06-28 00:00:00.001', '2020-06-28 00:00:00.001', '2020-06-28 00:00:00.005'),
```
Comment (wluyima) [inline app-api/src/test/resources/mgt_sender_sync_message.sql:2]: I know this is test data, but in practice event_date should be older than date_sent

File: webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.component.html [hunk 17]
```diff
@@ -27,8 +28,9 @@ <h5><span class="badge badge-dark" i18n="@@common-total-count">Total Count</span
 			<i class="bi-envelope-dash text-warning ui-cursor-pointer" [hidden]="m.status != 'NEW'" [ngbTooltip]="statusWaitingTemplate" [openDelay]="1000"></i>
 			<i class="bi-envelope-check-fill text-success ui-cursor-pointer" [hidden]="m.status != 'SENT'" [ngbTooltip]="statusSentTemplate" [openDelay]="1000"></i>
 		</td>
-		<td>{{m.dateCreated | date:'yyyy-MM-dd HH:mm:ss'}}</td>
-		<td>{{m.dateSent | date:'yyyy-MM-dd HH:mm:ss'}}</td>
+		<td container="body" ngbTooltip="{{m.eventDate | date:'EEEE, MMMM d, y, h:mm:ss'}}" [openDelay]="1000">{{m.eventDate | date:'MMM d, y'}}</td>
```
Comment (wluyima) [inline webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.component.html:31]: date format should be `mediumDate` and tooltip date `EEEE, MMMM d, y, h:mm:ss a`

Comment (danielchirinda) [reply webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.component.html:31]: That is the format of mediumDate 'MMM d, y'

File: app-api/src/test/java/org/openmrs/eip/app/route/sender/DbSyncRouteTest.java (hunk 15 above)
Comment (danielchirinda) [reply app-api/src/test/java/org/openmrs/eip/app/route/sender/DbSyncRouteTest.java:60]: Without calling getTime method i had this result: 
java.lang.AssertionError: expected:<2022-10-07 19:34:08.758> but was:<Fri Oct 07 19:34:08 CAT 2022>
	at org.junit.Assert.fail(Assert.java:89)
	.....
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)

File: webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.component.html (hunk 17 above)
Comment (wluyima) [reply webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.component.html:31]: I know but lets use `dd` for date for all dates

Comment (wluyima) [reply webapp/src/main/webapp/src/app/sender/sync/sender-sync-message.component.html:31]: And for the tooltip to use dd and include AM/PM, this is more intuitive to users

File: app-api/src/test/java/org/openmrs/eip/app/route/sender/DbSyncRouteTest.java (hunk 15 above)
Comment (wluyima) [reply app-api/src/test/java/org/openmrs/eip/app/route/sender/DbSyncRouteTest.java:60]: Ok, then that's fine, I think I have run into this before too, one of them does not have the nano seconds

Comment (wluyima) [reply app-api/src/test/java/org/openmrs/eip/app/route/sender/DbSyncRouteTest.java:60]: Ok, I think I have run into this before, I wonder why one excludes the nano seconds

Comment (wluyima) [reply app-api/src/test/java/org/openmrs/eip/app/route/sender/DbSyncRouteTest.java:60]: I think it is a problem with junit with how it compares the date objects when they have nano seconds
